[CheckpointWrapper] Reentrant kwarg support

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #85062
* __->__ #84908
* #84907

A temporary patch to support keyword args when reentrant checkpoint wrapper is used. This is need to unblock some crucial workloads, the ideal fix would be checking this directly into torch.utils.checkpoint. 

Differential Revision: [D39453453](https://our.internmc.facebook.com/intern/diff/D39453453/)

rohan-varma committed:
[CheckpointWrapper] Reentrant kwarg support

Per title

Differential Revision: [D39453453](https://our.internmc.facebook.com/intern/diff/D39453453/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[CheckpointWrapper] Reentrant kwarg support"


Per title

Differential Revision: [D39453453](https://our.internmc.facebook.com/intern/diff/D39453453/)

[ghstack-poisoned]

rohan-varma commented on torch/distributed/algorithms/_checkpoint/checkpoint_wrapper.py:
```diff
@@ -76,11 +76,38 @@ def forward(self, *args, **kwargs):
+            # Support keyword arguments for reentrant checkpoint. Note that this
+            # only works if user has specified self.checkpoint_impl and is not
+            # using their own custom checkpoint_fn.
+            if self.checkpoint_impl == CheckpointImpl.REENTRANT and kwargs != {}:
```
debating whether we need the `kwargs != {}` or if we should just do this anyways even if there are no kwargs.

awgu replied:
I think the `kwargs != {}` is fine. Otherwise, we construct a new function at runtime unnecessarily (among some other unnecessary ops). Is there any reason to use this code path even when `kwargs` is empty?

rohan-varma committed:
Update on "[CheckpointWrapper] Reentrant kwarg support"


A temporary patch to support keyword args when reentrant checkpoint wrapper is used. This is need to unblock some crucial workloads, the ideal fix would be checking this directly into torch.utils.checkpoint. 

Differential Revision: [D39453453](https://our.internmc.facebook.com/intern/diff/D39453453/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[CheckpointWrapper] Reentrant kwarg support"


A temporary patch to support keyword args when reentrant checkpoint wrapper is used. This is need to unblock some crucial workloads, the ideal fix would be checking this directly into torch.utils.checkpoint. 

Differential Revision: [D39453453](https://our.internmc.facebook.com/intern/diff/D39453453/)

[ghstack-poisoned]

awgu commented on torch/distributed/utils.py:
```diff
@@ -4,10 +4,40 @@
 
+def _pack_kwargs(*args: Any, **kwargs: Any) -> Tuple[Tuple[str, ...], Tuple[Any, ...]]:
+    """
+    Turn argument list into separate key list and value list (unpack_kwargs does the opposite)
```
I realize this is a private function, but could we consider adding a section to the docstring like:
```
Returns:
    Tuple[Tuple[str, ...], Tuple[Any, ...]]: The first tuple element gives
    the kwarg keys, and the second tuple element gives both positional args
    and kwarg values, where the positional args proceed kwarg values and
    kwarg values are ordered consistently with the kwarg keys. The first tuple
    element's length is at most the second tuple element's length.
```
My motivation for this is to document the implicit assumptions that the `_unpack_kwargs` method has on the `_pack_kwargs` return value.

For Python 3.7+, the order of `dict` iteration is guaranteed to be the insertion order, so it may be okay to not specify that the returned `tuple` may not match the _original_ kwargs order (if `kwargs.items()` did not guarantee order). The important thing is that the order between the 1st and 2nd returned `tuple` elements are consistent.

rohan-varma replied:
Thanks for the suggestion, added this docstring!

awgu commented on torch/distributed/utils.py:
```diff
@@ -4,10 +4,40 @@
+    Turn argument list into separate key list and value list (unpack_kwargs does the opposite)
+    Inspiration: https://github.com/facebookresearch/fairscale/blob/eeb6684/fairscale/internal/containers.py#L70
+    Usage::
+
```
I wonder if this API design may be slightly confusing due to the inversion of `tuple` ordering.
- From Python, `*args` naturally proceeds `*kwargs`.
- `_pack_kwargs` returns `kwarg_keys` as the first element and `flat_args` as the second element.
- `_unpack_kwargs` takes `kwarg_keys` as the first arg and `flat_args` as the second arg. This nicely matches the return value of `_pack_kwargs`.
- `_unpack_kwargs` returns `args` as the first element and `kwargs` as the second element.

Specifically, I am wondering if we should reorder the `_pack_kwargs` return value to return `flat_args, kwarg_keys` instead. This might be more natural, but this is totally up to you. If you do invert the return value order, then the type annotation and sample docstring above need to be adjusted as well.

rohan-varma replied:
I see, so we want the intermediate to be consistent with the overall. I think that makes sense.

rohan-varma committed:
Update on "[CheckpointWrapper] Reentrant kwarg support"


A temporary patch to support keyword args when reentrant checkpoint wrapper is used. This is need to unblock some crucial workloads, the ideal fix would be checking this directly into torch.utils.checkpoint. 

Differential Revision: [D39453453](https://our.internmc.facebook.com/intern/diff/D39453453/)

[ghstack-poisoned]

awgu commented on torch/distributed/utils.py:
```diff
@@ -4,10 +4,40 @@
     _is_namedtuple
 )
-from typing import Dict, Any, List
+from typing import Dict, Any, List, Tuple
```
nit: sort?

awgu commented on torch/distributed/algorithms/_checkpoint/checkpoint_wrapper.py:
```diff
@@ -148,16 +175,20 @@ def checkpoint_wrapper(
-            offloaded to CPU. Default is ``False``.
+            offloaded to CPU. Default is ``False``. Wrappers with activation
+            offload can be composed with ones that do recomputation-based
+            checkpoint to trade off increased compute versus CPU memory usage
```
> trade off increased compute versus CPU memory usage and H2D transfer.

Clarification:
- Activation checkpointing means _increased_ compute.
- Activation offload means _increased_ CPU memory usage and _additional_ H2D transfers.

If this is correct, then can we consider adding:
```suggestion
            checkpoint to trade off increased compute versus increased CPU memory usage
            and additional H2D transfers.
```
(just to be explicit on the directionality of effects in the tradeoff)

awgu reviewed (approved):
LGTM. I left some comments, but none are blocking.

rohan-varma committed:
Update on "[CheckpointWrapper] Reentrant kwarg support"


A temporary patch to support keyword args when reentrant checkpoint wrapper is used. This is need to unblock some crucial workloads, the ideal fix would be checking this directly into torch.utils.checkpoint. 

Differential Revision: [D39453453](https://our.internmc.facebook.com/intern/diff/D39453453/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[CheckpointWrapper] Reentrant kwarg support"


A temporary patch to support keyword args when reentrant checkpoint wrapper is used. This is need to unblock some crucial workloads, the ideal fix would be checking this directly into torch.utils.checkpoint. 

Differential Revision: [D39453453](https://our.internmc.facebook.com/intern/diff/D39453453/)

[ghstack-poisoned]

rohan-varma commented on test/distributed/fsdp/test_checkpoint_wrapper.py:
```diff
@@ -56,6 +56,58 @@ def test_load_activation_checkpointed_module(self):
+                    self.assertEqual(t1, t2)
+                    self.assertEqual(t1.requires_grad, t2.requires_grad)
+
+        # Test model that enforces kwarg inputs
```
This just means that the model forward function enforces kwargs, because its signature is:

```
def forward(self, *, a=None, b=None)
```

so all other arguments than `self` have to be kwargs. Test is mostly here because it was a request from FLAVA team to make sure that this works.

rohan-varma commented on torch/distributed/algorithms/_checkpoint/checkpoint_wrapper.py:
```diff
@@ -76,11 +76,38 @@ def forward(self, *args, **kwargs):
-            )
+            # Support keyword arguments for reentrant checkpoint. Note that this
+            # only works if user has specified self.checkpoint_impl and is not
+            # using their own custom checkpoint_fn.
```
Filed https://github.com/pytorch/pytorch/issues/85025 to track this, but the high level is:

1. checkpoint_fn is needed because some use cases (internal) want to use a checkpoint function other than torch.utils.checkpoint with the wrapper, in this case we'll ignore checkpoint_impl and use the passed in function.
2. offload_to_cpu=True is definitely not well designed at the moment, we should revamp.

rohan-varma committed:
Update on "[CheckpointWrapper] Reentrant kwarg support"


A temporary patch to support keyword args when reentrant checkpoint wrapper is used. This is need to unblock some crucial workloads, the ideal fix would be checking this directly into torch.utils.checkpoint. 

Differential Revision: [D39453453](https://our.internmc.facebook.com/intern/diff/D39453453/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[CheckpointWrapper] Reentrant kwarg support"


A temporary patch to support keyword args when reentrant checkpoint wrapper is used. This is need to unblock some crucial workloads, the ideal fix would be checking this directly into torch.utils.checkpoint. 

Differential Revision: [D39453453](https://our.internmc.facebook.com/intern/diff/D39453453/)

[ghstack-poisoned]

rohan-varma commented:
@pytorchbot merge -g
